### ISSUE TITLE
constant number of elements in paginator

### ISSUE THREAD

**Issue by KasperJanssens** [pbrisbin/yesod-paginator#26]
Hello,

Right now there is a pageCount which, if I understand correctly, functions by taking at most pageCount elements left and at most pageCount elements right of the active page. This is slightly annoying as it makes my layout act differently in the middle of the paginator (where ca 'pageCount' * 2 elements are in the paginator)  and at the ends (where there is 'pageCount' number of elements in the paginator). All of a sudden those links show larger or smaller depending on where I am. This does not seem like something.

Was there a reason for this? I can see that right now the code is rather simple and elegant, with taking 'pageCount' of elements before and after, and with a constant number of pages there could be more corner cases to be handled of course

**Comment by pbrisbin** [comment]
Hey there, thanks for the report. It's been a long time since I've worked on this project, so I'm not sure I can speak to the original intentions or what would or wouldn't make sense to implement or support -- but I'm happy to explore your issue in general.

Can you describe a little more how you'd want it to work? What should it look like when you're on the first or last page in your ideal UI?

**Comment by KasperJanssens** [comment]
Well, let's say that I want to limit it to four pages and in total there are twenty pages. When the first page is active I would like to see 1,2,3,4 and the dots and the arrows and so on. When on the last page I would like to see 17, 18, 19 and 20, and the dots and the arrows and so on. When I'm on page 10 I would like to see 8, 9, 10 and 11 for example (or replace 8 by 12, fine as well). The point is I would like to have the same amount of UI elements in the paginator regardless of on which page I am. If that makes sense to you let me know :P. If it does not, the same, I'll try to elaborate a bit (although it is a rather primitive idea right now, but the point I am coming from is that I would like to have a constant amount of UI elements in the paginator, for styling reasons).

**Comment by pbrisbin** [comment]
OK, I think I see. For 20 pages, `pageCount = 4`, you want it work like this:

One page 1:

```
(1) [2] [3] [4] ... [20] >
```

On page 20:

```
< [1] ... [17] [18] [19] (20) >
```

On page 10:

```
< [1] ... [8] [9] (10) [11] ... [20]  >

or

< [1] ... [9] (10) [11] [12] ... [20] >
```

And if it looked this, that'd still be no good right?

```
< [1] ... [8] [9] (10) [11] [12] ... [20] >
```

**Comment by KasperJanssens** [comment]
Yes, that's indeed what I was saying and I see now that this is not completely ok yet. 
I would prefer a constant number of links in the paginator. 

The first and second example have 5 page links, 1 ellipsis and two arrows. 
The third and fourth example unfortunately has 6 page links, two ellipsis and two arrows. Hmm, especially those ellipsis are annoying as they will not always both be present.

What would be possible is this, let's say with `pageCount = 8` instead of 4. So, on page 1:
`< (1) [2] [3] [4] [5] [6] ... [20] >`

On page 20:
`< [1] ... [15] [16] [17] [18] [19] (20)>`

On page 5:

`< [1] ... [4] (5) [6] [7] ... [20] >`

Does that make more sense? PageCount is undoubtedly an extremely bad name in the end for this type of config. But it means the amount of visible pagelinks on the paginator.

**Comment by pbrisbin** [comment]
To be honest, I think the trouble is the ellipsis and first/last elements. I just looked at Google's pagination links as a quick example and they don't have either of those:

![](https://files.pbrisbin.com/screenshots/screenshot.24735.png)

I feel like we could also drop them and meet your requirement while keeping things simple:

Page 1:

```
< (1) [2] [3] [4] [5] [6] [7] [8] >
```

Page 20:

```
< [13] [14] [15] [16] [17] [18] [19] (20) >
```

Page 10:

```
< [6] [7] [8] [9] (10) [11] [12] [13] >
```

Would you be willing to give up ellipsis and first/last to achieve the consistent element count?

**Comment by KasperJanssens** [comment]
Hmm, checking wat Google does, good call. Sure, all that really matters to me is having a constant amount of ui elements, and I never noticed Google not having first and last, so that means I never missed it neither. Fine by me!

**Comment by pbrisbin** [comment]
I went ahead and did some re-architecting  on this project in #24. I'm really happy with the result: it simplifies a lot (it only looks like a big-addition PR because I added a lot of docs, tests, and filled out the CHANGELOG too), makes things way more flexible, and I implemented the preferred widget as the simple one discussed here.

So this Issue should close pretty soon, once I get that version released. Bear in mind it'll be a major version bump and the interface is different:

https://github.com/pbrisbin/yesod-paginator/blob/00a276fa3b56f25c2445da10a5b3d5df88e853cd/src/Yesod/Paginator.hs#L7-L18

**Comment by KasperJanssens** [comment]
Great, thanks! I'll guess I'll best leave the issue open until the release has been done?

**Comment by pbrisbin** [comment]
Yup, that seems best to me.

**Comment by pbrisbin** [comment]
Released as `v1.1.0.0` :tada:
